File: static/js/timer.js
'use strict';

export class Timer extends HTMLSpanElement {
	constructor() {
		super();
		this.time = 0;
	}
	connectedCallback() {
		this.innerText = this.display;
	}
	get display() {
		const minutes = Math.floor(this.time / 60);
		const seconds = this.time % 60;
		let minText = minutes >= 10 ? minutes.toString() : '0' + minutes.toString();
		let secText = seconds >= 10 ? seconds.toString() : '0' + seconds.toString();
		return `${minText}:${secText}`;
	}
	get running() {
		return Boolean(this.interval);
	}
	update(time) {
		this.time = time;
		this.innerText = this.display;
	}
	resume() {
		if (!this.running) {
			this.interval = setInterval(() => {
				if (this.time > 0) {
					this.time -= 1;
					this.innerText = this.display;
				} else this.pause();
			}, 1000);
		}
	}
	pause() {
		clearInterval(this.interval);
		this.interval = null;
	}
}

File: src/timer.rs
use std::time::{Duration, Instant};

pub struct Timer {
    time: Duration,
    mark: Instant,
    running: bool,
}

impl Timer {
    pub fn new(time: Duration, running: bool) -> Timer {
        Timer {
            time,
            mark: Instant::now(),
            running
        }
    }
    pub fn time(&self) -> Duration {
        if self.running {
            self.time.saturating_sub(
                Instant::now().duration_since(self.mark)
            )
        } else {
            self.time
        }
    }
    pub fn running(&self) -> bool {
        self.running
    }
    pub fn resume(&mut self) {
        if !self.running {
            self.mark = Instant::now();
            self.running = true;
        }
    }
    pub fn pause(&mut self) {
        if self.running {
            self.time = self.time();
            self.running = false;
        }
    }
    pub fn add(&mut self, time: Duration) {
        self.time += time;
    }
}
